src/main/java/io/jenkins/plugins/tuleap_api/client/internals/TuleapAuthenticationApiAuthenticationClient.java

```diff
@@ -0,0 +1,166 @@
+import okhttp3.*;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.logging.Logger;
+
+public class TuleapAuthenticationApiAuthenticationClient implements AccessTokenApi, OpenIDClientApi {
+
+    public static final Logger LOGGER = Logger.getLogger(TuleapAuthenticationApiAuthenticationClient.class.getName());
+
+    private final PluginHelper pluginHelper;
+    private final OkHttpClient client;
+    private final ObjectMapper objectMapper;
+    private final UrlJwkProvider jwkProvider;
+    private final TuleapConfiguration tuleapConfiguration;
+    private final HeaderAuthenticationValidator headerAuthenticationValidator;
+    private final AccessTokenValidator accessTokenValidator;
+    private final UserInfoValidator userInfoValidator;
+
+    @Inject
+    public TuleapAuthenticationApiAuthenticationClient(
+        PluginHelper pluginHelper,
+        OkHttpClient client,
+        ObjectMapper objectMapper,
+        UrlJwkProvider jwkProvider,
+        TuleapConfiguration tuleapConfiguration,
+        HeaderAuthenticationValidator headerAuthenticationValidator,
+        AccessTokenValidator accessTokenValidator,
+        UserInfoValidator userInfoValidator
+    ) {
+        this.pluginHelper = pluginHelper;
+        this.client = client;
+        this.objectMapper = objectMapper;
+        this.jwkProvider = jwkProvider;
+        this.tuleapConfiguration = tuleapConfiguration;
+        this.headerAuthenticationValidator = headerAuthenticationValidator;
+        this.accessTokenValidator = accessTokenValidator;
+        this.userInfoValidator = userInfoValidator;
+    }
+
+    @Override
+    @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE") // see https://github.com/spotbugs/spotbugs/issues/651
+    public AccessToken getAccessToken(
+        String codeVerifier,
+        String authorizationCode,
+        String clientId,
+        Secret clientSecret
+    ) {
+        Request accessTokenRequest = this.buildAccessTokenRequestBody(authorizationCode, codeVerifier, clientId, clientSecret);
+
+        try (Response response = this.client.newCall(accessTokenRequest).execute()) {
+            if (!response.isSuccessful()) {
+                throw new InvalidTuleapResponseException(response);
+            }
+            this.headerAuthenticationValidator.validateHeader(response);
+            this.accessTokenValidator.validateAccessTokenHeader(response);
+            AccessToken accessToken = this.objectMapper.readValue(Objects.requireNonNull(response.body()).string(), AccessTokenEntity.class);
+            this.accessTokenValidator.validateAccessTokenBody(accessToken);
+            return accessToken;
+        } catch (IOException | InvalidTuleapResponseException | InvalidHeaderException exception) {
+            LOGGER.severe(exception.getMessage());
+            throw new RuntimeException("Error while contacting Tuleap server", exception);
```

Comment:
We could discuss this tomorrow. I think you could make them directly RuntimeExceptions